BUILD_SHARED_LIB=ON → disaster

I'm trying to build dReal with `BUILD_SHARED_LIB=ON`. It fails _horribly_. Rather than try to include the myriad link errors, here are some brief examples:

```
linking lib/libdop.so (cc)
CMakeFiles/dop.dir/tools/dop/print_latex.cpp.o: In function `operator<<':
.../dreal/src/opensmt/egraph/Enode.h:380: undefined reference to `Enode::print(std::ostream&) const'
CMakeFiles/dop.dir/tools/dop/process.cpp.o: In function `operator()<__gnu_cxx::__normal_iterator<std::pair<std::basic_string<char>, Enode*>*, std::vector<std::pair<std::basic_string<char>, Enode*> > >, __gnu_cxx::__normal_iterator<std::pair<std::basic_string<char>, Enode*>*, std::vector<std::pair<std::basic_string<char>, Enode*> > > >':
.../dreal/src/tools/dop/process.cpp:175: undefined reference to `dreal::starts_with(std::string const&, std::string const&)'
```

```
.../lib/libdreal.so: undefined reference to `filib::fp_traits_base<double>::nan_val'
.../lib/libdreal.so: undefined reference to `ibex::IntervalVector::mid() const'
.../lib/libdreal.so: undefined reference to `glp_ipt_status'
```

This is happening on multiple machines (so far, on Fedora 21 and Ubuntu 14.04, but I suspect the problem is universal), as of @e6d1d49.

CMake arguments:
```
USE_NLOPT=OFF
BUILD_TESTING=OFF
BUILD_SHARED_LIB=ON
CMAKE_SHARED_LINKER_FLAGS=-Wl,--no-undefined
```

Possibly related to #51. In particular, https://github.com/dreal/dreal3/issues/51#issuecomment-89334447 suggests that this is a known issue that for some reason was not fixed, despite that #51 was closed.

--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -811,6 +811,17 @@ else()
   add_library(dop STATIC ${DOP_SRCS})
 endif()
 
+if(BUILD_SHARED_LIBS)
+  set(CMAKE_SHARED_LINKER_FLAGS "${CMAKE_SHARED_LINKER_FLAGS} ${DREAL_EXTRA_LINKER_FLAGS} ${EXTERNAL_LD_FLAGS}")
+
+  target_link_libraries(dreal ${EXTRA_LIBS})
+  if(${PYTHONLIBS_FOUND})
+    target_link_libraries(dop dreal ${PYTHON_LIBRARIES} ${EXTRA_LIBS})
+  else()
+    target_link_libraries(dop dreal ${EXTRA_LIBS})
+  endif()
+endif()
+
 add_dependencies(dop PICOSAT)
 add_dependencies(dop IBEX)
 add_dependencies(dop ADEPT)

